components/raftstore/src/store/peer.rs

```diff
@@ -3696,16 +3704,72 @@ where
 
         let transfer_leader = get_transfer_leader_cmd(&req).unwrap();
         let peer = transfer_leader.get_peer();
-
-        let transferred = self.pre_transfer_leader(peer);
-
+        let transferred = if self.peer.id == peer.id {
+            true
+        } else {
+            match self.propose_locks_before_transfer_leader(ctx) {
+                Ok(true) => self.propose_normal(ctx, req).is_ok(),
+                Ok(false) => self.pre_transfer_leader(peer),
+                Err(_) => false,
+            }
+        };
         // transfer leader command doesn't need to replicate log and apply, so we
         // return immediately. Note that this command may fail, we can view it just as an advice
         cb.invoke_with_response(make_transfer_leader_response());
 
         transferred
     }
 
+    // Returns whether it needs to propose the transfer leader command instead of
+    // just sending messages.
+    fn propose_locks_before_transfer_leader<T>(
+        &mut self,
+        ctx: &mut PollContext<EK, ER, T>,
+    ) -> Result<bool> {
+        // 1. Disable in-memory pessimistic locks
+        let mut pessimistic_locks = self.txn_ext.pessimistic_locks.write();
+        let locks = if pessimistic_locks.valid {
+            // FIXME: `valid` should be set back to true if the peer remains being leader after
+            // some time (which means leader transfer fails)
+            pessimistic_locks.valid = false;
+            // If there is no lock when setting `valid` to false, we can guarantee that no lock
+            // needs to be migrated to the new leader. In this case, we can just send a message
+            // to trigger leader transfer instead of making a proposal.
+            if pessimistic_locks.map.is_empty() {
+                return Ok(false);
+            }
+            mem::take(&mut pessimistic_locks.map)
+        } else {
+            // If `valid` is false, the locks may have been proposed but not committed.
+            // So, it is not safe to use a Raft message to transfer leader in case the lock
+            // proposal fails.
+            return Ok(true);
+        };
+        drop(pessimistic_locks);
+
+        // 2. Propose pessimistic locks
```

Comment:
What if there are some raft logs that will remove in-memory pessimistic locks queued before these requests? Can these requests bring locks back?

Reply:
Yes, it will be incorrect...

I'm thinking if removing the in-memory pessimistic locks in the `proposed_cb` is enough.

Reply:
Well, the current `proposed_cb` may be not suitable because it's not guaranteed to be called...

Reply:
It's not easy to solve. Let us think about it.

Reply:
I change the way of writing in-memory pessimistic locks in https://github.com/tikv/tikv/pull/11577/commits/72184155ecf838c6b435d4ddafe3486788622803. I think removing the locks before proposing should work.

Reply:
I have the same idea at the beginning, but I'm thinking it may lose some memory locks if fails to propose these requests that remove `removed_pessimistic_locks` due to leadership transfer. After digging into it, I think the possibility is very low because there is pre-transfer-leader.

Reply:
Yes... Locks may be lost if the scheduler thread runs much slower than the raftstore, then the write request may be proposed after transfer leader. The possibility is low, but it's good to avoid it.